### PR TITLE
Refactor provider init

### PR DESCRIPTION
**Pull request by cpuguy83**
- Refactor provider init
i. Moves provider init out of vkubelet setup, instead preferring to initialize vkubelet with a provider.
- Split API server configuration from setup.
i. This makes sure that configuration (which is done primarily through env vars) is separate from actually standing up the servers
ii. This also makes sure to abort daemon initialization if the API servers are not able to start

### CODE REVIEW

File: cmd/root.go [hunk 1]
```diff
@@ -50,11 +66,21 @@ var RootCmd = &cobra.Command{
...
+			MetricsAddr:     metricsAddr,
+			Provider:        p,
+			ResourceManager: rm,
+			APIConfig:       apiConfig,
+		})
 		if err != nil {
 			log.L.WithError(err).Fatal("Error initializing virtual kubelet")
 		}
-		if err := f.Run(context.Background()); err != nil {
+		if err := f.Run(ctx); err != nil {
```
Comment (robbiezhang) [inline cmd/root.go]: will it fix #363

File: cmd/root.go [hunk 2]
```diff
@@ -155,4 +181,61 @@ func initConfig() {
 	})
 	logger.Level = level
 	log.L = logger
+
+	if !disableTaint {
+		taint, err = getTaint(taintKey, provider)
+		if err != nil {
+			logger.WithError(err).Fatal("Error setting up desired kubernetes node tain")
```
Comment (robbiezhang) [inline cmd/root.go]: /tain/taint

File: vkubelet/vkubelet.go [hunk 3]
```diff
@@ -189,7 +138,6 @@ func (s *Server) registerNode(ctx context.Context) error {
 				Architecture:    "amd64",
 				KubeletVersion:  "v1.11.2",
 			},
-			Capacity:        s.provider.Capacity(ctx),
```
Comment (robbiezhang) [inline vkubelet/vkubelet.go]: why Capacity is not set?

Comment (cpuguy83) [reply vkubelet/vkubelet.go]: Good question. Don't remember even touching this. Nice catch.

File: cmd/client.go [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+		// Get the kubeconfig from the filepath.
+		config, err = clientcmd.BuildConfigFromFlags("", configPath)
+		if err != nil {
+			return nil, errors.Wrap(err, "error building client config")
+		}
+	} else {
+		// Set to in-cluster config.
+		config, err = rest.InClusterConfig()
+		if err != nil {
+			return nil, errors.Wrap(err, "error building in cliuster config")
```
Comment (robbiezhang) [inline cmd/client.go]: /cliuster/cluster/

File: providers/register/register.go [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+	ResourceManager *manager.ResourceManager
+}
+
+type initFunc func(InitConfig) (providers.Provider, error)
+
+// Get gets the provider specified by the given name
+func Get(name string, cfg InitConfig) (providers.Provider, error) {
+	f, ok := providerInits[name]
+	if !ok {
+		return nil, strongerrors.NotFound(errors.New("provider not found"))
```
Comment (robbiezhang) [inline providers/register/register.go]: include the provider name in the message

File: vkubelet/vkubelet.go [hunk 6]
```diff
@@ -32,123 +31,73 @@ type Server struct {
...
-		}
-	} else {
-		// Set to in-cluster config.
-		config, err = rest.InClusterConfig()
-		if err != nil {
-			return nil, err
-		}
-	}
+type APIConfig struct {
+	Cert string
```
Comment (robbiezhang) [inline vkubelet/vkubelet.go]: CertPath?

File: vkubelet/vkubelet.go [hunk 7]
```diff
@@ -32,123 +31,73 @@ type Server struct {
...
+	Key  string
```
Comment (robbiezhang) [inline vkubelet/vkubelet.go]: KeyPath?

File: vkubelet/vkubelet.go [hunk 8]
```diff
@@ -32,123 +31,73 @@ type Server struct {
...
-	}
+	ctx = log.WithLogger(ctx, log.G(ctx))
 
-	rm, err := manager.NewResourceManager(clientset)
+	apiL, err := net.Listen("tcp", cfg.APIConfig.Addr)
 	if err != nil {
-		return nil, pkgerrors.Wrap(err, "error creating resource manager")
+		return nil, pkgerrors.Wrap(err, "error setting up API listener")
 	}
+	defer func() {
```
Comment (robbiezhang) [inline vkubelet/vkubelet.go]: this seems wrong. the apiL should not be closed in this method.

File: vkubelet/vkubelet.go [hunk 9]
```diff
@@ -32,123 +31,73 @@ type Server struct {
...
-	var defaultTaintValue string
-	if taintKey != "" {
-		defaultTaintKey = taintKey
-		defaultTaintValue = ""
+	if cfg.MetricsAddr != "" {
+		metricsL, err := net.Listen("tcp", cfg.MetricsAddr)
+		if err != nil {
+			return nil, pkgerrors.Wrap(err, "error setting up metrics listener")
+		}
+		defer func() {
```
Comment (robbiezhang) [inline vkubelet/vkubelet.go]: same here. metricsL should not be closed in the function

File: vkubelet/vkubelet.go (hunk 8 above)
Comment (cpuguy83) [reply vkubelet/vkubelet.go]: This is only closing if the main function returns an error.

File: providers/register/register.go [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+	OperatingSystem string
+	InternalIP      string
+	DaemonPort      int32
+	ResourceManager *manager.ResourceManager
+}
+
+type initFunc func(InitConfig) (providers.Provider, error)
+
+// Get gets the provider specified by the given name
+func Get(name string, cfg InitConfig) (providers.Provider, error) {
```
Comment (robbiezhang) [inline providers/register/register.go]: change the function name to GetProvider?

File: cmd/root.go (hunk 1 above)
Comment (cpuguy83) [reply cmd/root.go]: I don't think so. Wasn't aware of the problem. Will take a look after this.

**Comment by cpuguy83** [comment]
This is all updated.
@jeremyrickard Could you take a look?

File: cmd/root.go [hunk 11]
```diff
@@ -16,18 +16,29 @@ package cmd
...
 import (
 	"context"
+	"fmt"
 	"os"
 	"path/filepath"
+	"strconv"
 	"strings"
 
 	"github.com/Sirupsen/logrus"
+	"github.com/cpuguy83/strongerrors"
```
Comment (jeremyrickard) [inline cmd/root.go]: this library is cool

**Pull request merged**